cirq-core/cirq/ops/control_values.py

```diff
@@ -38,104 +39,99 @@ class AbstractControlValues(abc.ABC):
+        """
 
     @abc.abstractmethod
-    def _identifier(self) -> Any:
-        """Returns the internal representation of the object."""
+    def _num_qubits_(self) -> int:
+        """Returns the number of qubits for which control values are stored by this object."""
 
     @abc.abstractmethod
-    def __hash__(self) -> int:
-        pass
+    def _json_dict_(self) -> Dict[str, Any]:
+        """Returns a dictionary used for serializing this object."""
 
     @abc.abstractmethod
-    def __repr__(self) -> str:
-        pass
+    def _circuit_diagram_info_(
+        self, args: 'cirq.CircuitDiagramInfoArgs'
+    ) -> 'cirq.CircuitDiagramInfo':
+        """Returns information used to draw this object in circuit diagrams."""
 
     @abc.abstractmethod
-    def validate(self, qid_shapes: Union[Tuple[int, ...], List[int]]) -> None:
-        """Validates control values
-
-        Validate that control values are in the half closed interval
-        [0, qid_shapes) for each qubit.
+    def __iter__(self) -> Iterator[Tuple[int, ...]]:
+        """Iterator on internal representation of control values used by the derived classes.
+
+        Note: Be careful that the terms iterated upon by this iterator will have different
+        meaning based on the implementation. For example:
+        >>> print(*cirq.ProductOfSums([(0, 1), (0,)]))
+        (0, 1) (0,)
+        >>> print(*cirq.SumOfProducts([(0, 0), (1, 0)]))
+        (0, 0) (1, 0)
         """
 
-    @abc.abstractmethod
-    def _are_ones(self) -> bool:
-        """Checks whether all control values are equal to 1."""
-
-    @abc.abstractmethod
-    def _json_dict_(self) -> Dict[str, Any]:
-        pass
-
-    def __iter__(self) -> Generator[Tuple[int, ...], None, None]:
-        for assignment in self._expand():
-            yield assignment
+    def _value_equality_values_(self) -> Any:
+        return tuple(v for v in self.expand())
 
-    def __eq__(self, other) -> bool:
-        """Returns True iff self and other represent the same configurations.
+    def __and__(self, other: 'AbstractControlValues') -> 'AbstractControlValues':
+        """Sets self to be the cartesian product of all combinations in self x other.
 
         Args:
-            other: A AbstractControlValues object.
+          other: An object that implements AbstractControlValues.
 
         Returns:
-            boolean whether the two objects are equivalent or not.
+          An object that represents the cartesian product of the two inputs.
         """
-        if not isinstance(other, AbstractControlValues):
-            other = ProductOfSums(other)
-        return sorted(v for v in self) == sorted(v for v in other)
+        return SumOfProducts(
+            tuple(x + y for (x, y) in itertools.product(self.expand(), other.expand()))
+        )
 
 
-@dataclass(frozen=True, eq=False)
 class ProductOfSums(AbstractControlValues):
     """ProductOfSums represents control values in a form of a cartesian product of tuples."""
```

Comment:
I think this would be much more understandable if we described the _meaning_ of this representation, rather than what types are used in the internal representation. This could also go into the `AbstractControlValues` docstring to explain what we mean by "product" and "sum", something like:

> Control values define predicates on the state of one or more qubits. Predicates can be composed with logical OR to form a "sum", or with logical AND to form a "product". We provide two implementations: `SumOfProducts` which consists of one or more AND (product) clauses each of which applies to all N qubits, and `ProductOfSums` which consists of N OR (sum) clauses, each of which applies to one qubit.

Reply:
The suggest docstring looks great! Updated docstring for the base class and both derived classes. PTAL.